pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1242,59 +1261,76 @@ func findBridge(bridgeName string) (*netlink.Bridge, error) {
 }
 
 // XXX Need additional logic for IPv6 underlays.
-func getUlAddrs(ctx *zedrouterContext,
-	ifnum int, appNum int,
-	status *types.UnderlayNetworkStatus,
-	netInstStatus *types.NetworkInstanceStatus) (string, string, error) {
-
-	log.Functionf("getUlAddrs(%d/%d)\n", ifnum, appNum)
+func getUlAddrs(ctx *zedrouterContext, ifnum int,
+	netInstStatus *types.NetworkInstanceStatus,
+	ulStatus *types.UnderlayNetworkStatus, appID uuid.UUID) (string, error) {
 
-	bridgeIPAddr := ""
-	appIPAddr := ""
-
-	// Allocate bridgeIPAddr based on BridgeMac
-	log.Functionf("getUlAddrs(%d/%d for %s) bridgeMac %s\n",
-		ifnum, appNum, netInstStatus.UUID.String(),
-		status.BridgeMac.String())
 	var err error
-	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
+	var mac net.HardwareAddr
+	ipAddr := ""
+	networkID := netInstStatus.UUID
+	if netInstStatus.Subnet.IP == nil ||
+		netInstStatus.DhcpRange.Start == nil {
+		log.Functionf("getUlAddrs(%d/%s) no subnet %s\n",
+			ifnum, networkID.String(), appID.String())
+		return ipAddr, nil
+	}
+	log.Functionf("getUlAddrs(%d) for %s/%s\n", ifnum,
+		networkID.String(), appID.String())
+	// XXX or change type of VifInfo.Mac to avoid parsing?
+	if ulStatus.Mac == "" {
+		log.Functionf("getUlAddrs(%d/%s) no mac %s\n",
+			ifnum, networkID.String(), appID.String())
+		return ipAddr, nil
+	}
+	mac, err = net.ParseMAC(ulStatus.Mac)
 	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
-	} else {
-		bridgeIPAddr = addr
-	}
-
-	if status.AppIPAddr != nil {
-		// Static IP assignment case.
-		// Note that appIPAddr can be in a different subnet.
-		// Assumption is that the config specifies a gateway/router
-		// in the same subnet as the static address.
-		appIPAddr = status.AppIPAddr.String()
-		recordIPAssignment(ctx, netInstStatus, status.AppIPAddr,
-			status.Mac)
-	} else if status.Mac != "" {
-		// XXX or change type of VifInfo.Mac to avoid parsing?
-		var mac net.HardwareAddr
-		mac, err = net.ParseMAC(status.Mac)
-		if err != nil {
-			log.Fatal("ParseMAC failed: ", status.Mac, err)
+		errStr := fmt.Sprintf("getUlAddrs(%s/%s): parse Mac fail: %v\n",
+			networkID.String(), appID.String(), err)
+		log.Errorf(errStr)
+		return ipAddr, errors.New(errStr)
+	}
+
+	// for static IP Address
+	if ulStatus.AppIPAddr != nil {
+		ipAddr = ulStatus.AppIPAddr.String()
+		if netInstStatus.DhcpRange.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("getUlAddrs(%s/%s): Dhcp Range contains the static IP Address %s\n",
+				networkID.String(), appID.String(), ipAddr)
+			log.Errorf(errStr)
+			return "", errors.New(errStr)
+		}
+	} else if netInstStatus.DhcpRange.Start != nil {
+		appNum, err := appNumOnUNetAlloc(ctx, networkID, appID,
```

Comment:
The condition for calling appNumOnUNetAlloc and appNumOnUNetFree/Delete do not seem to match. Thus I suspect there are cases when you might free/delete when there was no alloc, and also the reverse.

I think it is much simpler to always call appNumOnUNetAlloc for every UlConfig when we handle the create, and then always free/delete when we inactive/delete, independent of the use of IP addresses. dhcprange, etc, etc.

Reply:
The app number is now, only specific to dynamically IP Address Range(Dhcp Range). We are not usning them for static IP Address and Bridge IP Address (, as they have defined IP Address, and also out of scope for the Bitmap). Please suggest, whether this is really needed, we are not using the app number for them.

Reply:
Have made the appNumOnUNetAlloc generic for everyone on the underlay network. And also made same for the delete/deactivate path for appNumOnUNetDelete/Free.